app/src/main/java/com/mapbox/maps/testapp/examples/terrain/SantaCatalinaActivity.kt

```diff
@@ -0,0 +1,311 @@
+    super.onCreate(savedInstanceState)
+    requestWindowFeature(Window.FEATURE_NO_TITLE)
+    setContentView(R.layout.activity_simple_map)
+    mapView.disablePlugins()
+
+    // get map and setup initial camera
+    mapboxMap = mapView.getMapboxMap()
+    mapboxMap.jumpTo(
+      CameraOptions.Builder()
+        .center(POINT_START)
+        .zoom(14.0)
+        .bearing(215.0)
+        .build()
+    )
+
+    // load satellite style and add terrain with a line layer to visualize the route
+    mapboxMap.loadStyle(
+      style(styleUri = Style.SATELLITE_STREETS) {
+        +rasterDemSource(SOURCE) {
+          url(TERRAIN_URL_TILE_RESOURCE)
+          tileSize(512)
+        }
+        +terrain(SOURCE) {
+          exaggeration(1.7)
+        }
+        +layerAtPosition(
+          lineLayer(LINE_LAYER_ID, GEOJSON_SOURCE_ID) {
+            lineColor(rgb(255, 79, 60))
+            lineWidth(5.0)
+          },
+          above = LAYER_ABOVE_ID
+        )
+        +locationIndicatorLayer(LOCATION_LAYER_ID) {
+          topImage(FOREGROUND_ICON)
+          bearingImage(BACKGROUND_ICON)
+          imagePitchDisplacement(5.0)
+          topImageSize(1.5)
+          bearingImageSize(1.5)
+        }
+        +image(FOREGROUND_ICON) {
+          bitmap(BitmapFactory.decodeResource(resources, R.drawable.mapbox_mylocation_icon_default))
+        }
+        +image(BACKGROUND_ICON) {
+          bitmap(
+            generateBitmap(
+              this@SantaCatalinaActivity,
+              R.drawable.mapbox_mylocation_bg_shape,
+              null
+            )
+          )
+        }
+      }
+    ) { style ->
+      // hide road labels
+      style.getLayerAs<SymbolLayer>(LAYER_ROAD_ID).visibility(Visibility.NONE)
+
+      // wait for map to become idle before making request
+      mapboxMap.addOnMapChangedListener {
+        if (it == MapChange.DID_BECOME_IDLE) {
+          // execute direction request
+          executeDirectionsRequestForRoute {
+            // add a source to visualize the route
+            addSourceForRoute(style, it)
+            // animate the route
+            animateRoute(it)
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Executes a directions request.
+   * propagates the linestring geometry as high order function.
+   */
+  private fun executeDirectionsRequestForRoute(animateRoute: (LineString) -> Unit) {
+    val client = MapboxDirections.builder()
+      .origin(POINT_START)
+      .destination(POINT_END)
+      .overview(DirectionsCriteria.OVERVIEW_SIMPLIFIED)
+      .profile(DirectionsCriteria.PROFILE_WALKING)
+      .steps(true)
+      .accessToken(getString(R.string.mapbox_access_token))
+      .build()
+    client.enqueueCall(object : Callback<DirectionsResponse> {
+      override fun onResponse(
+        call: Call<DirectionsResponse>,
+        response: Response<DirectionsResponse>
+      ) {
+        animateRoute(
+          LineString.fromPolyline(
+            response.body()!!.routes()[0].geometry()!!,
```

Comment:
nit: we could avoid using `!!` to handle errors more gracefully.

Reply:
I will optimize for it but I disagree that we should always replace `!!` with graceful error handling. In the SDK, we should never use `!!`. In the test app we are allowed to use it. My personal believe is that it's better to fail fast and you at least know something is wrong. Versus silently ignoring and printing out logcat message.

Reply:
for context I'm replacing it with this code:

```
        response.body()?.let { body ->
          body.routes()[0].geometry()?.let {
            animateRoute(
              LineString.fromPolyline(it, PRECISION_6)
            )
            return
          }
        }
        throw RuntimeException("Not able to retrieve a directions route")
```